### PR TITLE
[CALCITE-5855] Implement frame exclusion in window functions

### PR DESCRIPTION
**Pull request by itiels**
This PR adds support for frame exclusion clause in window functions.

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/rel/logical/ToLogicalConverterTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/rel/logical/ToLogicalConverterTest.java:485]: missing empty line?
do you want to test the other kinds of exclusions too?

File: core/src/main/java/org/apache/calcite/rel/core/Window.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/rel/core/Window.java:368]: wouldn't the "no ties" also apply here?

**Review by mihaibudiu — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/rex/RexBuilder.java
Comment (asolimando) [reply core/src/main/java/org/apache/calcite/rex/RexBuilder.java:392]: The current policy is exactly what @mihaibudiu suggests, you should leave the old API around with a default value for the new parameter (exclude no others?), at that point no need to document in `history.md` IMO.

By doing this, the number of files to be updated in the scope of this PR will drop.

If you want to deprecate the old API, you can do so with the `@Deprecated` annotation (see [SqlDrop.java#L37](https://github.com/apache/calcite/blob/3c4d54a82e21ec6926e6aa724ce94ead63d19083/core/src/main/java/org/apache/calcite/sql/SqlDrop.java#L37) for an example), but I don't think it's necessary here.

**Review by asolimando — CHANGES_REQUESTED** [review]
Please leave the old API as suggested, in order to avoid unnecessary breaking changes.

File: core/src/main/java/org/apache/calcite/rex/RexBuilder.java
Comment (itiels) [reply core/src/main/java/org/apache/calcite/rex/RexBuilder.java:392]: Thanks for reviewing @mihaibudiu and @asolimando . 
So should I add overload option for each public method that I've changed? And how do I decide if the old versions should be deprecated or not?

File: core/src/main/java/org/apache/calcite/rel/core/Window.java
Comment (itiels) [reply core/src/main/java/org/apache/calcite/rel/core/Window.java:368]: You are correct, great catch.

File: core/src/main/java/org/apache/calcite/rex/RexBuilder.java
Comment (asolimando) [reply core/src/main/java/org/apache/calcite/rex/RexBuilder.java:392]: From a cursory look, the SQL parser still accepts the same syntax (without `exclude` statement) after your modifications.

If that's confirmed, then I think there is no need to deprecate the old APIs, as they would basically mimic the same (legal) old behaviour.

In a nut-shell, you are adding an optional functionality, what was working before should still be working as-is, it's not a change in behaviour.

Comment (itiels) [reply core/src/main/java/org/apache/calcite/rex/RexBuilder.java:413]: Ok. I've pushed the changes, please let me know if that is what you had in mind.

File: core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java
Comment (asolimando) [inline core/src/main/java/org/apache/calcite/rel/logical/LogicalWindow.java:70]: In the codebase we generally don't format javadoc as you do here, can you please revert these changes?

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java
Comment (asolimando) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:968]: Nit: why not a regular case and throwing on `default`? 

I am also wondering if we want to omit printing the exclusion when there is no exclusion, it would keep the generated SQL backward compatible to what we have been generating now. WDYT?

File: core/src/main/java/org/apache/calcite/rex/RexWindowExclusion.java
Comment (asolimando) [inline core/src/main/java/org/apache/calcite/rex/RexWindowExclusion.java:50]: Nit: what about a switch similar to what you wrote for `SqlImplementor`?

File: core/src/test/resources/sql/winagg.iq
Comment (asolimando) [inline core/src/test/resources/sql/winagg.iq:851]: Can you also add the same line but without "exclude no others" to verify they results in the same resultset? (here and in similar added tests)

File: site/_docs/algebra.md
Comment (asolimando) [inline site/_docs/algebra.md:545]: Nit: can you remove the extra spaces at the end of these lines?

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
Comment (asolimando) [inline testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:5784]: The comment is probably not needed as we have the `ok` method just after

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
Comment (asolimando) [inline testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:5791]: Why do we need the `'^'` here?

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
Comment (asolimando) [inline testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:5798]: Please add a newline here

**Review by asolimando — CHANGES_REQUESTED** [review]
LGTM, some more comments but all minor

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:5791]: @asolimando this is a convention used to indicate the error position reported by the parser in the statement

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java
Comment (itiels) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:968]: I've replace this assert with exception. 
Regarding generated SQL, we do ignore the exclude in case of `Exclude no others` in `SqlWindow.unparse`.

File: core/src/main/java/org/apache/calcite/rex/RexWindowExclusion.java
Comment (itiels) [reply core/src/main/java/org/apache/calcite/rex/RexWindowExclusion.java:50]: Maybe I'm missing something, but I think I can't do that here because I have `SqlLiteral` and not an Enum.

**Comment by itiels** [comment]
Thanks again @asolimando! 
I've fixed according to your requests, and fixed some of the sonar cloud issues (some were indeed out of scope).

**Review by asolimando — APPROVED** [review]
LGTM, @itiels thanks for taking care of the comments, since @mihaibudiu already approved I think you can optimistically squash the commits for preparing to merge the PR (please make sure the only commit message aligns with the Jira title etc.). 

I generally wait one or more extra days before merging to give time for other people to react.

**Review by mihaibudiu — APPROVED** [review]

File: site/_docs/history.md
Comment (NobiGo) [inline site/_docs/history.md:52]: This file is a template file, we don't need to change this. The release manager will modify it.

File: site/_docs/algebra.md
Comment (NobiGo) [inline site/_docs/algebra.md:533]: This change is necessary?

File: core/src/test/resources/sql/winagg.iq
Comment (NobiGo) [inline core/src/test/resources/sql/winagg.iq:844]: PostgreSQL may be better.

Comment (NobiGo) [reply core/src/test/resources/sql/winagg.iq:844]: Adding the Jira summary is better.

**Review by NobiGo — CHANGES_REQUESTED** [review]
Need some change to confirm.

**Comment by itiels** [comment]
Thanks for the review @NobiGo , I've fixes the issues. 
@asolimando I've squashed the commits. Thank you.

**Comment by mihaibudiu** [comment]
@NobiGo if you are satisfied we can merge this

**Pull request merged**
